src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.start).before(dist)
+
+    def after(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.stop - 1).after(dist)
+
+    def prev(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes leading up to the start?
+        raise NotImplementedError('Block.prev')
+
+    def next(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes past the end?
+        raise NotImplementedError('Block.next')
+
+    # ------------------------------------------------------------------------ #
+    # Sequence interface implementation
+    # ------------------------------------------------------------------------ #
+
+    def __iter__(self):
+        attr, _range = self._path[-1]
+        block = self.parent()
+        for i in _range:
+            yield block.child(attr, i)
+
+    def __getitem__(self, i):
+        attr, r = self._path[-1]
+        r = r[i]
+        if isinstance(r, range):
+            if r.step != 1:
+                raise IndexError('block cursors must be contiguous')
+            return Block(self._proc, self._path[:-1] + [(attr, r)])
+        else:
+            return self.parent().child(attr, r)
+
+    def __len__(self):
+        _, _range = self._path[-1]
+        return len(_range)
+
+    # ------------------------------------------------------------------------ #
+    # AST mutation
+    # ------------------------------------------------------------------------ #
+
+    def _replace(self, nodes: list, *, empty_default=None):
+        """
+        This is an UNSAFE internal function for replacing a block in an AST with
+        a list of statements and providing a forwarding function as collateral.
+        It is meant to be package-private, not class-private, so it may be
+        called from other internal classes and modules, but not from end-user
+        code.
+        """
+        assert self._path
+        assert len(self) > 0
+
+        def update(node):
+            attr, i = self._path[-1]
+            children = getattr(node, attr)
+            new_children = children[:i.start] + nodes + children[i.stop:]
+            new_children = new_children or empty_default or []
+            return node.update(**{attr: new_children})
+
+        from .API import Procedure
+        p = Procedure(self._rewrite_node(update))
+
+        return p, self._forward_replace(weakref.ref(p), len(nodes))
+
+    def _forward_replace(self, new_proc, n_ins):
+        del_range = self._path[-1][1]
```

Comment:
Does
`_, del_range = self._path[-1]`
work?  If so, it might be more consistent with your style elsewhere in this file.

Reply:
It does, and will fix.